File: 이하영/240519/4949. 균형잡힌 세상/균형잡힌 세상.py
while True:
    str = input()
    stack = []
    
    if str == '.':
        break
    
    for i in str:
        if i == '[' or i == '(':
            stack.append(i)
        elif i == ']':
            if len(stack) != 0 and stack[-1] == '[':  # stack이 비워있지 않거나 이 전에 '['일 경우
                stack.pop()  # 스택에서 지워줌
            else:
                stack.append(i)  #그게 아니라면 stack에 추가하고 종료
                break
        elif i == ')':
            if len(stack) != 0 and stack[-1] == '(':  # stack이 비워있지 않거나 이 전에 '('일 경우
                stack.pop()  # 스택에서 지워줌
            else:
                stack.append(i)  #그게 아니라면 stack에 추가하고 종료
                break
    
    if len(stack) == 0:
        print("yes")
    else:
        print("no")

File: 한민선/240520/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        Stack<Character> stack;

        while(true){
            String str = br.readLine();
            if(str.charAt(0) == '.' && str.length() == 1){
                break;
            }
            stack = new Stack<>();
            for (int i = 0; i < str.length(); i++) {
                if(str.charAt(i) == '(' || str.charAt(i) == '['){
                    stack.push(str.charAt(i));
                }
                if(!stack.isEmpty()){
                    if(str.charAt(i) == ')') {
                        if(stack.peek() == '('){
                            stack.pop();
                        }else {
                            stack.push(str.charAt(i));
                        }
                    } else if(str.charAt(i) == ']'){
                        if(stack.peek() == '['){
                            stack.pop();
                        }else {
                            stack.push(str.charAt(i));
                        }
                    }
                }else{
                    if(str.charAt(i) == ')' || str.charAt(i) == ']'){
                        stack.push(str.charAt(i));
                    }
                }
            }
            if(stack.isEmpty()){
                bw.write("yes \n");
            }else{
                bw.write("no \n");
            }
        }

        bw.flush();
        bw.close();
    }
}